### PR TITLE
prune-junit-xml: join stdout when merging tests

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The intent in https://github.com/kubernetes/kubernetes/pull/127929 was to strip log messages out of the JUnit failure text and then keep the complete text as stdout. That didn't work when merging different tests at the package level because that step dropped the stdout of each test.

##### With current master and unstructured multi-line log entries, stripping out the log messages leaves additional lines behind:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1850686612854280192

```
k8s.io/kubernetes/test/integration: examples expand_less 	7m31s
{Failed;Failed;  === RUN   TestFrontProxyConfig/WithUID
    testserver.go:582: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:402: runtime-config=map[api/all:true]
    testserver.go:403: Starting kube-apiserver on port 39919...
    testserver.go:438: Waiting for /healthz to be ok...
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
...
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.authentication.k8s.io v1alpha1.coordination.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.authentication.k8s.io v1beta1.networking.k8s.io v1beta1.storage.k8s.io v1beta3.flowcontrol.apiserver.k8s.io v2.autoscaling]
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2024-10-28 00:38:08.456636897 +0000 UTC m=+377.378006483"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "system" and it exists`,
  }
...
```

##### When converting that log call to structured logging, the extra diff gets filtered out properly:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129035/pull-kubernetes-integration/1862558341121708032

##### Smarter merging of texts and preserving the full stdout text

See commit messages for details.

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129035/pull-kubernetes-integration/1863500189038284800

```
k8s.io/kubernetes/test/integration: examples
{Failed  === RUN   TestFrontProxyConfig/WithoutUID
...
--- FAIL: TestFrontProxyConfig/WithoutUID (65.73s)

=== RUN   TestFrontProxyConfig/WithUID
    testserver.go:582: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
...
=== RUN   TestFrontProxyConfig
--- FAIL: TestFrontProxyConfig (131.02s)
}
open stdout
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by pohly** [comment]
/test pull-kubernetes-verify

At the moment the PR intentionally fails in some integration tests and only runs a few of them, to demonstrate the problem. The fix will come later.

**Comment by pohly** [comment]
I've removed the commits which I used to trigger failures. This PR is now ready for review.

/assign @dims @aojea

**Comment by pohly** [comment]
/assign @Rajalakshmi-Girish

Can you perhaps review? You implemented the pruning of tests initially.

**Review by krzyzacy — APPROVED** [review]

**Review by krzyzacy — APPROVED** [review]
/lgtm

File: cmd/prune-junit-xml/prunexml.go [hunk 1]
```diff
@@ -135,34 +135,54 @@ func pruneTESTS(suites *junitxml.JUnitTestSuites) {
...
 func joinTexts(a, b string) string {
 	if a == "" {
 		return b
 	}
 	if b == "" {
 		return a
 	}
-	return a + "; " + b
+	if strings.Contains(a, b) {
+		return a
```
Comment (Rajalakshmi-Girish) [inline cmd/prune-junit-xml/prunexml.go:179]: The `joinTexts` function is responsible for appending subtest failure details to create a consolidated list.

Each subtest must retain all its details individually. If the function fails to append the details when they are already present, it can result in missing information.

Is there a way to display all the details of a single subtest at once before proceeding to the next? This approach would ensure clarity during the aggregation process.

Please see the below snapshots where the `=== RUN   TestFrontProxyConfig/WithUID` is missing it's failure message.

image
image

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

File: cmd/prune-junit-xml/prunexml.go (hunk 1 above)
Comment (pohly) [reply cmd/prune-junit-xml/prunexml.go:179]: > If the function fails to append the details when they are already present, it can result in missing information.

If they are already present, they cannot be missing... Saying the same thing twice without some meta context doesn't add information.

> Please see the below snapshots where the `=== RUN   TestFrontProxyConfig/WithUID` is missing it's failure message.

This is from https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129035/pull-kubernetes-integration/1863500189038284800, right?

What you want didn't work before either. This used to be rendered as:
```
{Failed; Failed; Failed  === RUN   TestFrontProxyConfig/WithoutUID ... 
...
--- FAIL: TestFrontProxyConfig/WithoutUID (64.89s)
; === RUN   TestFrontProxyConfig/WithUID
```

In other words, the "Failed" summary and the detailed failure messages are detached.

It would be possible to put the test name before the summary and each failure message. I think that would just make the entire output longer without making it more useful. It's already visible what failed.

**Review by Rajalakshmi-Girish — APPROVED** [review]
/lgtm

File: cmd/prune-junit-xml/prunexml.go (hunk 1 above)
Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml.go:179]: Although this approach looks good to me, there’s a potential risk of mismatches between the count of 'summaries' and 'detailed failure messages.'
This mismatch could lead to ambiguity or difficulty in mapping summaries to their corresponding test failures, especially if the summaries provide crucial information beyond just stating that a test 'failed.'

Please pardon my ignorance if I’m overstating a minor concern.

Comment (pohly) [reply cmd/prune-junit-xml/prunexml.go:179]: The JUnit file produced by gotestsum always has "Failed" as failure message, and those are the only files that we process with prune-junit-xml.
